.gitignore

```diff
@@ -10,3 +10,12 @@
 *.out
 *.lock
 *.doit*
+
+# nbsite
+# these files normally shouldn't be checked in as they should be
+# dynamically built from notebooks
+doc/**/*.rst
+doc/**/*.ipynb
+doc/**/*.json
```

Comment:
Does ** represent a special gitignore syntax, compared to *?

Reply:
My understanding is that * denotes 1 level of nesting whereas ** denotes arbitrary levels of nesting. It's just glob syntax:

```
(pyviz) ➜  pyviz git:(master) ✗ ls doc/*/*.rst
doc/topics/index.rst                              doc/tutorial/08_Custom_Interactivity.rst
doc/topics/nyc_taxi.rst                           doc/tutorial/09_Operations_and_Pipelines.rst
doc/tutorial/00_Setup.rst                         doc/tutorial/10_Working_with_Large_Datasets.rst
doc/tutorial/00_Welcome.rst                       doc/tutorial/11_Streaming_Data.rst
doc/tutorial/01_Workflow_Introduction.rst         doc/tutorial/12_Parameters_and_Widgets.rst
doc/tutorial/02_Annotating_Data.rst               doc/tutorial/13_Deploying_Bokeh_Apps.rst
doc/tutorial/03_Customizing_Visual_Appearance.rst doc/tutorial/A1_Exploration_with_Containers.rst
doc/tutorial/04_Working_with_Tabular_Data.rst     doc/tutorial/A2_Dashboard_Workflow.rst
doc/tutorial/05_Working_with_Gridded_Data.rst     doc/tutorial/index.rst
doc/tutorial/06_Network_Graphs.rst                doc/tutorial/scipy18.rst
doc/tutorial/07_Geographic_Data.rst               doc/tutorial/strata_ny18.rst
(pyviz) ➜  pyviz git:(master) ✗ ls doc/**/*.rst
doc/FAQ.rst                                       doc/tutorial/05_Working_with_Gridded_Data.rst
doc/Roadmap.rst                                   doc/tutorial/06_Network_Graphs.rst
doc/about.rst                                     doc/tutorial/07_Geographic_Data.rst
doc/background.rst                                doc/tutorial/08_Custom_Interactivity.rst
doc/index.rst                                     doc/tutorial/09_Operations_and_Pipelines.rst
doc/installation.rst                              doc/tutorial/10_Working_with_Large_Datasets.rst
doc/topics/index.rst                              doc/tutorial/11_Streaming_Data.rst
doc/topics/nyc_taxi.rst                           doc/tutorial/12_Parameters_and_Widgets.rst
doc/tutorial/00_Setup.rst                         doc/tutorial/13_Deploying_Bokeh_Apps.rst
doc/tutorial/00_Welcome.rst                       doc/tutorial/A1_Exploration_with_Containers.rst
doc/tutorial/01_Workflow_Introduction.rst         doc/tutorial/A2_Dashboard_Workflow.rst
doc/tutorial/02_Annotating_Data.rst               doc/tutorial/index.rst
doc/tutorial/03_Customizing_Visual_Appearance.rst doc/tutorial/scipy18.rst
doc/tutorial/04_Working_with_Tabular_Data.rst     doc/tutorial/strata_ny18.rst
```

Reply:
I'm confused.  Do you have a special shell?  Bash doesn't do anything different for ** that I can see:

```
bash-3.2$ ls doc/*/*.rst
doc/tutorial/00_Setup.rst				doc/tutorial/10_Working_with_Large_Datasets.rst
doc/tutorial/00_Welcome.rst				doc/tutorial/11_Streaming_Data.rst
doc/tutorial/01_Workflow_Introduction.rst		doc/tutorial/12_Parameters_and_Widgets.rst
doc/tutorial/02_Annotating_Data.rst			doc/tutorial/13_Deploying_Bokeh_Apps.rst
doc/tutorial/03_Customizing_Visual_Appearance.rst	doc/tutorial/2018_PyViz_Quansight_webinar.rst
doc/tutorial/04_Working_with_Tabular_Data.rst		doc/tutorial/A1_Exploration_with_Containers.rst
doc/tutorial/05_Working_with_Gridded_Data.rst		doc/tutorial/A2_Dashboard_Workflow.rst
doc/tutorial/06_Network_Graphs.rst			doc/tutorial/index.rst
doc/tutorial/07_Geographic_Data.rst			doc/tutorial/scipy18.rst
doc/tutorial/08_Custom_Interactivity.rst		doc/tutorial/strata_ny18.rst
doc/tutorial/09_Operations_and_Pipelines.rst
bash-3.2$ ls doc/**/*.rst
doc/tutorial/00_Setup.rst				doc/tutorial/10_Working_with_Large_Datasets.rst
doc/tutorial/00_Welcome.rst				doc/tutorial/11_Streaming_Data.rst
doc/tutorial/01_Workflow_Introduction.rst		doc/tutorial/12_Parameters_and_Widgets.rst
doc/tutorial/02_Annotating_Data.rst			doc/tutorial/13_Deploying_Bokeh_Apps.rst
doc/tutorial/03_Customizing_Visual_Appearance.rst	doc/tutorial/2018_PyViz_Quansight_webinar.rst
doc/tutorial/04_Working_with_Tabular_Data.rst		doc/tutorial/A1_Exploration_with_Containers.rst
doc/tutorial/05_Working_with_Gridded_Data.rst		doc/tutorial/A2_Dashboard_Workflow.rst
doc/tutorial/06_Network_Graphs.rst			doc/tutorial/index.rst
doc/tutorial/07_Geographic_Data.rst			doc/tutorial/scipy18.rst
doc/tutorial/08_Custom_Interactivity.rst		doc/tutorial/strata_ny18.rst
doc/tutorial/09_Operations_and_Pipelines.rst
```

Reply:
I see from https://docs.python.org/3/library/glob.html `Changed in version 3.5: Support for recursive globs using “**”.`  So it's recent for Python, and I see that git does support it (sorry for the noise!), but I'm surprised that you see different shell behavior than I do.